src/gurobi_optimods/metromap.py

```diff
@@ -0,0 +1,1242 @@
+    linepaths : DataFrame
+        DataFrame with information on the line routes/paths.
+        This data frame could also be empty
+    include_planarity : bool
+        parameter to turn off the consideration of planarity constraints
+    geodata : bool
+        if original positions were given
+    penalty_edge_directions : float
+        weight for original direction part in the objective
+    penalty_line_bends : float
+        weight for line-bend part in the objective
+    penalty_distance : float
+        weight for distance part in the objective
+
+    Returns
+    -------
+    tuple
+        - networkx graph with node property pos_oct representing the x and y coordinates of the octilinear representation
+        - Python dict for direction for each edge (needed for plotting)
+    """
+
+    edge_directions_out = dict()
+    with create_env() as env, gp.Model(env=env) as model:
+        num_nodes = len(graph.nodes)
+        mindist = 1  # required distance in x- or y-coordinate for each edge
+        # variables for position (coordinates)
+        x = model.addVars(graph.nodes, ub=num_nodes, name="x-coord")
+        y = model.addVars(graph.nodes, ub=num_nodes, name="y-coord")
+        dist = {}
+        for u, v in graph.edges:
+            dist[u, v] = model.addVar(name=f"distance{u},{v}")
+            dist[v, u] = model.addVar(name=f"distance{v},{u}")
+            model.addConstr(dist[u, v] == dist[v, u])
+            model.addConstr(dist[u, v] >= x[v] - x[u] - 1)
+            model.addConstr(dist[u, v] >= y[v] - y[u] - 1)
+            model.addConstr(dist[u, v] >= x[u] - x[v] - 1)
+            model.addConstr(dist[u, v] >= y[u] - y[v] - 1)
+
+        # direction for each edge (octilinear)
+        edge_direction = {}
+        for u, v in graph.edges:
+            for i in range(8):
+                edge_direction[u, v, i] = model.addVar(
+                    vtype="B", name=f"direction[{u},{v},{i}]"
+                )
+                edge_direction[v, u, i] = model.addVar(
+                    vtype="B", name=f"direction[{v},{u},{i}]"
+                )
+            # symmetric of back and forth direction (need to wait until all variables for (u,v) are created)
+            model.addConstr(edge_direction[u, v, 0] == edge_direction[v, u, 4])
+            model.addConstr(edge_direction[u, v, 1] == edge_direction[v, u, 5])
+            model.addConstr(edge_direction[u, v, 2] == edge_direction[v, u, 6])
+            model.addConstr(edge_direction[u, v, 3] == edge_direction[v, u, 7])
+            model.addConstr(edge_direction[u, v, 4] == edge_direction[v, u, 0])
+            model.addConstr(edge_direction[u, v, 5] == edge_direction[v, u, 1])
+            model.addConstr(edge_direction[u, v, 6] == edge_direction[v, u, 2])
+            model.addConstr(edge_direction[u, v, 7] == edge_direction[v, u, 3])
+
+        # objective edge length
+        obj = [0]
```

Comment:
Does `obj` have to be a list? I can only see you using `obj[0]`

Reply:
You are right. I used this because then the obj was changed in the functions. But probably, the correct way is to return obj in the functions where it is changed. I updated accordingly.